### PR TITLE
Add DMS Serverless Operators

### LINKED ISSUE #39954: Adding DMS serverless options to aws provider
### Description

Sometime back AWS [launched](https://aws.amazon.com/blogs/aws/new-aws-dms-serverless-automatically-provisions-and-scales-capacity-for-migration-and-data-replication/) serverless DMS (Data Migration Service). This lets users create replication jobs without explicitly setting up a replication instance.

Currently, in the AWS provider, we have support for classic DMS, which requires users to set up a [replication instance](https://github.com/prakharcode/airflow/blob/93ba263732a30ed51fdae0b2bff7cd959be115b8/airflow/providers/amazon/aws/operators/dms.py#L78). It would be nice, if we can provide dms serverless support using the following boto3 functions:

- [create_replication_config](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dms/client/create_replication_config.html)
- [describe-replication-configs](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/dms/describe-replication-configs.html)
- [describe-replications](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/dms/describe-replications.html)
- [start-replication](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/dms/start-replication.html)
- [stop-replication](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/dms/stop-replication.html)
- [delete-replication-config](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/dms/delete-replication-config.html)

### Use case/motivation

Giving airflow[aws] a serverless DMS option would unlock specific DMS use cases that are becoming increasingly common via serverless, like -- full-load of a new table and patching a CDC table using the full load on serverless. 

The serverless implementation is also quite appealing for source database replication that happens at a larger frequency than a couple of days (where a dedicated instance doesn't make sense) or with source data that changes or grows a lot.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ellisms**
---
Adding operators, waiters, and triggers to support AWS DMS Serverless replications.
closes #39954

### CODE REVIEW

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py [hunk 1]
```diff
@@ -219,3 +224,158 @@ def wait_for_task_status(self, replication_task_arn: str, status: DmsTaskWaiterS
...
+        :return: List of replication tasks
+        """
+        filters = filters if filters is not None else []
+
+        try:
+            resp = self.conn.describe_replication_configs(Filters=filters, **kwargs)
+            return resp.get("ReplicationConfigs", [])
+        except Exception as ex:
+            self.log.error("Error while describing replication configs: %s", str(ex))
+            return []
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/hooks/dms.py]: Do you really want to swallow any exception and return an empty list in that case? We might miss some issues

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py [hunk 2]
```diff
@@ -219,3 +224,158 @@ def wait_for_task_status(self, replication_task_arn: str, status: DmsTaskWaiterS
...
+
+        :param filters: List of filter objects
+        :return: List of replications
+        """
+        filters = filters if filters is not None else []
+        try:
+            resp = self.conn.describe_replications(Filters=filters, **kwargs)
+            return resp.get("Replications", [])
+        except Exception:
+            return []
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/hooks/dms.py]: Same

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py [hunk 3]
```diff
@@ -219,3 +224,158 @@ def wait_for_task_status(self, replication_task_arn: str, status: DmsTaskWaiterS
             ],
             WithoutSettings=True,
         )
+
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/hooks/dms.py:227]: Overall I am wondering if it is worth creating all these methods, they are all just wrapper around boto3 apis. You added some error management around it but I am just questioning if it is worth creating new methods for that.

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 4]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+        self.deferrable = deferrable
+        self.waiter_delay = waiter_delay
+        self.waiter_max_attempts = waiter_max_attempts
+
+    def execute(self, context: Context) -> None:
+        results = self.hook.describe_replications(
+            filters=[{"Name": "replication-config-arn", "Values": [self.replication_config_arn]}]
+        )
+
+        if len(results) > 0:
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]: Do we need this? I expect `self.hook.get_provision_status` to fail if the replication does not exist? Which would make sense in that scenario

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 5]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+                and provision_status in self.TERMINAL_PROVISION_STATES
+            ):
+                self.log.info("DMS replication config(%s) is in valid state.", self.replication_config_arn)
+
+                self.hook.delete_replication_config(
+                    replication_config_arn=self.replication_config_arn,
+                    delay=self.waiter_delay,
+                    max_attempts=self.waiter_max_attempts,
+                )
+                self.handle_delete_wait()
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]: In order to simplify things I think you can remove that block of code. The `else` statement would also work if the replication config is in valid state

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 6]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+            self.log.info("Deferring until replication config is deleted.")
+            self.defer(
+                trigger=DmsReplicationConfigDeletedTrigger(
+                    replication_config_arn=self.replication_config_arn,
+                    waiter_delay=self.waiter_delay,
+                    waiter_max_attempts=self.waiter_max_attempts,
+                    aws_conn_id=self.aws_conn_id,
+                ),
+                method_name="execute_complete",
+            )
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]: That should be put under `if self.wait_for_completion:`. If `deferrable` is `True` but `wait_for_completion` is `False`, you should not wait

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 7]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+    def execute(self, context: Context):
+        result = self.hook.describe_replications(
+            filters=[{"Name": "replication-config-arn", "Values": [self.replication_config_arn]}]
+        )
+
+        try:
+            current_status = result[0].get("Status", "")
+        except Exception as ex:
+            self.log.error("Error while getting replication status: %s. Unable to start replication", str(ex))
+            raise ex
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]: Same here, I dont think it is necessary

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 8]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+                self.log.info("Deferring until deprovisioning completes.")
+                self.defer(
+                    trigger=DmsReplicationDeprovisionedTrigger(
+                        replication_config_arn=self.replication_config_arn,
+                        waiter_delay=self.waiter_delay,
+                        waiter_max_attempts=self.waiter_max_attempts,
+                        aws_conn_id=self.aws_conn_id,
+                    ),
+                    method_name="retry_execution",
+                )
```
Comment (vincbeck) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py:661]: This comment also apply for the above operator. I am wondering if you should not split this operator in two. One sensor would wait for the deprovisioning to complete, and this operator would fail if the status is not complete before actually doing it. I just think this operator does a lot and the user does not necessarily know that the operator could potentially wait for some time (with a cost) even though `wait_for_completion` is `False`

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py (hunk 1 above)
Comment (ellisms) [reply providers/src/airflow/providers/amazon/aws/hooks/dms.py]: Good point. I'll take another look at it.

File: providers/src/airflow/providers/amazon/aws/operators/dms.py (hunk 8 above)
Comment (ellisms) [reply providers/src/airflow/providers/amazon/aws/operators/dms.py:661]: Good point. The DMS docs weren't clear that a replication must be stopped (or failed) AND deprovisioned, so I was trying to cover both cases to simplify the deletion process. System testing pointed this out. It does add a bit to the operator, but the wait time would be the same in the Sensor + Operator combo. I see meeting the pre-requisite for deleting different than `wait_for_completion` of the actual delete. Open to exploring creating a sensor and would like to hear what others think.

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py (hunk 3 above)
Comment (ellisms) [reply providers/src/airflow/providers/amazon/aws/hooks/dms.py:227]: The `describe_*` methods get used in a few places, so I think those should be in the hook to avoid duplication. I'll look to consolidate some of the simpler ones directly into the Operator.

File: providers/src/airflow/providers/amazon/aws/operators/dms.py (hunk 5 above)
Comment (ellisms) [reply providers/src/airflow/providers/amazon/aws/operators/dms.py]: Which block of code?

File: docs/apache-airflow-providers-amazon/operators/dms.rst [hunk 9]
```diff
@@ -114,6 +114,69 @@ To delete a replication task you can use
...
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_dms_serverless_describe_replication]
+    :end-before: [END howto_operator_dms_serverless_describe_replication]
+
+Delete a serverless replication configuration
+=============================================
+
+To delete a serverless replication config use
+:class:`~airflow.providers.amazon.aws.operators.dms.DmsDescribeReplicationsOperator`.
```
Comment (o-nikolas) [inline docs/apache-airflow-providers-amazon/operators/dms.rst]: Copy/paste error?

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py (hunk 3 above)
Comment (o-nikolas) [reply providers/src/airflow/providers/amazon/aws/hooks/dms.py:227]: I think some below are quite simple, but others are decent number of lines that's nice to abstract out of the Operators (start and create for example). So I'm on the fence here, I could see both ways, but might lean on keeping it how it is since you've already spent time writing and testing the code as it is?

File: providers/src/airflow/providers/amazon/aws/triggers/dms.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        waiter_max_attempts: int = 60,
+        aws_conn_id: str | None = "aws_default",
+    ) -> None:
+        super().__init__(
+            serialized_fields={"replication_config_arn": replication_config_arn},
+            waiter_name="replication_complete",
+            waiter_delay=waiter_delay,
+            waiter_args={"Filters": [{"Name": "replication-config-arn", "Values": [replication_config_arn]}]},
+            waiter_max_attempts=waiter_max_attempts,
+            failure_message="Replication failed to reach terminal status.",
```
Comment (o-nikolas) [inline providers/src/airflow/providers/amazon/aws/triggers/dms.py]:
```suggestion
            failure_message="Replication failed to complete.",
```

Otherwise it's the same message as `DmsReplicationTerminalStatusTrigger` uses.

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 11]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+        Describe AWS DMS replication configurations.
+
+        :return: List of replication configurations
+        """
+        return self.hook.describe_replication_configs(filters=self.filter)
+
+
+class DmsCreateReplicationConfigOperator(AwsBaseOperator[DmsHook]):
+    """
+
```
Comment (o-nikolas) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]:
```suggestion
```

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 12]
```diff
@@ -277,3 +288,551 @@ def execute(self, context: Context):
...
+            additional_config_kwargs=self.additional_config_kwargs,
+        )
+
+        self.log.info("DMS replication config(%s) has been created.", self.replication_config_id)
+        return resp
+
+
+class DmsDeleteReplicationConfigOperator(AwsBaseOperator[DmsHook]):
+    """
+
```
Comment (o-nikolas) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]:
```suggestion
```

File: providers/src/airflow/providers/amazon/aws/operators/dms.py (hunk 5 above)
Comment (vincbeck) [reply providers/src/airflow/providers/amazon/aws/operators/dms.py]: The one I selected for this comment. From line `470`  to `483`

File: providers/src/airflow/providers/amazon/aws/hooks/dms.py (hunk 3 above)
Comment (ellisms) [reply providers/src/airflow/providers/amazon/aws/hooks/dms.py:227]: I'm inclined to leave it as-is since its already tested.

File: providers/tests/system/amazon/aws/example_dms_serverless.py [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+    DmsStartReplicationOperator,
+)
+from airflow.providers.amazon.aws.operators.rds import (
+    RdsCreateDbInstanceOperator,
+    RdsDeleteDbInstanceOperator,
+)
+from airflow.providers.amazon.aws.operators.s3 import S3CreateBucketOperator, S3DeleteBucketOperator
+from airflow.utils.trigger_rule import TriggerRule
+
+from tests.system.providers.amazon.aws.utils import ENV_ID_KEY, SystemTestContextBuilder
```
Comment (eladkal) [inline providers/tests/system/amazon/aws/example_dms_serverless.py]:
```suggestion
from providers.tests.system.amazon.aws.utils import ENV_ID_KEY, SystemTestContextBuilder
```

File: providers/tests/system/amazon/aws/example_dms_serverless.py [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+from tests.system.providers.amazon.aws.utils.ec2 import get_default_vpc_id
```
Comment (eladkal) [inline providers/tests/system/amazon/aws/example_dms_serverless.py]:
```suggestion
from providers.tests.system.amazon.aws.utils.ec2 import get_default_vpc_id
```

File: providers/src/airflow/providers/amazon/aws/operators/dms.py [hunk 15]
```diff
@@ -17,12 +17,23 @@
 # under the License.
 from __future__ import annotations
 
+from datetime import datetime
 from collections.abc import Sequence
```
Comment (eladkal) [inline providers/src/airflow/providers/amazon/aws/operators/dms.py]:
```suggestion
from collections.abc import Sequence
from datetime import datetime
```

File: docs/apache-airflow-providers-amazon/operators/dms.rst [hunk 16]
```diff
@@ -114,6 +114,69 @@ To delete a replication task you can use
...
     :end-before: [END howto_operator_dms_delete_task]
 
+
+Create a serverless replication config
+======================================
+
+To create a serverless replication config use
+:class:`~airflow.providers.amazon.aws.operators.dms.DmsCreateReplicationConfigOperator`.
+
+.. exampleinclude:: /../../tests/system/providers/amazon/aws/example_dms_serverless.py
```
Comment (eladkal) [inline docs/apache-airflow-providers-amazon/operators/dms.rst]:
```suggestion
.. exampleinclude:: /../../providers/tests/system/amazon/aws/example_dms_serverless.py
```

File: docs/apache-airflow-providers-amazon/operators/dms.rst [hunk 17]
```diff
@@ -114,6 +114,69 @@ To delete a replication task you can use
...
+    :start-after: [START howto_operator_dms_create_replication_config]
+    :end-before: [END howto_operator_dms_create_replication_config]
+
+Describe a serverless replication config
+========================================
+
+To describe a serverless replication config use
+:class:`~airflow.providers.amazon.aws.operators.dms.DmsDescribeReplicationConfigsOperator`.
+
+.. exampleinclude:: /../../tests/system/providers/amazon/aws/example_dms_serverless.py
```
Comment (eladkal) [inline docs/apache-airflow-providers-amazon/operators/dms.rst]:
```suggestion
.. exampleinclude:: /../../providers/tests/system/amazon/aws/example_dms_serverless.py
```

File: docs/apache-airflow-providers-amazon/operators/dms.rst [hunk 18]
```diff
@@ -114,6 +114,69 @@ To delete a replication task you can use
...
+    :start-after: [START howto_operator_dms_describe_replication_config]
+    :end-before: [END howto_operator_dms_describe_replication_config]
+
+Start a serverless replication
+==============================
+
+To start a serverless replication use
+:class:`~airflow.providers.amazon.aws.operators.dms.DmsStartReplicationOperator`.
+
+.. exampleinclude:: /../../tests/system/providers/amazon/aws/example_dms_serverless.py
```
Comment (eladkal) [inline docs/apache-airflow-providers-amazon/operators/dms.rst]:
```suggestion
.. exampleinclude:: /../../providers/tests/system/amazon/aws/example_dms_serverless.py
```

File: docs/apache-airflow-providers-amazon/operators/dms.rst [hunk 19]
```diff
@@ -114,6 +114,69 @@ To delete a replication task you can use
...
+    :start-after: [START howto_operator_dms_serverless_start_replication]
+    :end-before: [END howto_operator_dms_serverless_start_replication]
+
+Get the status of a serverless replication
+==========================================
+
+To get the status of a serverless replication use
+:class:`~airflow.providers.amazon.aws.operators.dms.DmsDescribeReplicationsOperator`.
+
+.. exampleinclude:: /../../tests/system/providers/amazon/aws/example_dms_serverless.py
```
Comment (eladkal) [inline docs/apache-airflow-providers-amazon/operators/dms.rst]:
```suggestion
.. exampleinclude:: /../../providers/tests/system/amazon/aws/example_dms_serverless.py
```

File: docs/apache-airflow-providers-amazon/operators/dms.rst [hunk 20]
```diff
@@ -114,6 +114,69 @@ To delete a replication task you can use
...
+    :start-after: [START howto_operator_dms_serverless_describe_replication]
+    :end-before: [END howto_operator_dms_serverless_describe_replication]
+
+Delete a serverless replication configuration
+=============================================
+
+To delete a serverless replication config use
+:class:`~airflow.providers.amazon.aws.operators.dms.DmsDeleteReplicationConfigOperator`.
+
+.. exampleinclude:: /../../tests/system/providers/amazon/aws/example_dms_serverless.py
```
Comment (eladkal) [inline docs/apache-airflow-providers-amazon/operators/dms.rst]:
```suggestion
.. exampleinclude:: /../../providers/tests/system/amazon/aws/example_dms_serverless.py
```

**Comment by eladkal** [comment]
The current error in the CI is a valid one:
`FAILED tests/always/test_project_structure.py::TestAmazonProviderProjectStructure::test_missing_examples - Failed: Not all classes are covered with example dags. Update self.MISSING_EXAMPLES_FOR_CLASSES if you want to skip this error`

You'll need to add example dags or exclude if there is a reason

**Comment by ellisms** [comment]
Added the missing example.

**Comment by eladkal** [comment]
Test fail :(

```
tests/always/test_example_dags.py::test_should_be_importable[providers/tests/system/amazon/aws/example_dms_serverless.py] - AssertionError: import_errors={'/opt/airflow/providers/tests/system/amazon/aws/example_dms_serverless.py': 'Traceback (most recent call last):\n  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed\n  File "/opt/airflow/providers/tests/system/amazon/aws/example_dms_serverless.py", line 472, in <module>\n    from tests.system.utils.watcher import watcher\nModuleNotFoundError: No module named \'tests.system.utils\'\n'}
assert 1 == 0
 +  where 1 = len({'/opt/airflow/providers/tests/system/amazon/aws/example_dms_serverless.py': 'Traceback (most recent call last):\n  Fi...e>\n    from tests.system.utils.watcher import watcher\nModuleNotFoundError: No module named \'tests.system.utils\'\n'})
 +    where {'/opt/airflow/providers/tests/system/amazon/aws/example_dms_serverless.py': 'Traceback (most recent call last):\n  Fi...e>\n    from tests.system.utils.watcher import watcher\nModuleNotFoundError: No module named \'tests.system.utils\'\n'} = <airflow.models.dagbag.DagBag object at 0x7f0814d88190>.import_errors

```

**Comment by ellisms** [comment]
The import error should be fixed now.

File: .pre-commit-config.yaml [hunk 21]
```diff
@@ -18,7 +18,7 @@
 default_stages: [pre-commit, pre-push]
 default_language_version:
   python: python3
-  node: 22.2.0
+  node: system #22.2.0
```
Comment (eladkal) [inline .pre-commit-config.yaml]: Is this change relevant to the PR?

Comment (ellisms) [reply .pre-commit-config.yaml]: No, I need to revert that.

**Review by eladkal — APPROVED** [review]
LGTM

**Comment by jscheffl** [comment]
It seems the merge of this PR broke tests on main/canary builds:
https://github.com/apache/airflow/actions/runs/12380031516/job/34555973915

Somebody having an idea how to fix? Shall we revert?

**Comment by jscheffl** [comment]
Okay, I thought I make a "quick fix" and add aiobotocore as dependency... but there is a bit of history and even as pre-commit check NOT to add this.
So loading/testing it seems need to be selective.

**Comment by potiuk** [comment]
Fix merged in #45013

File: providers/tests/system/amazon/aws/example_dms_serverless.py [hunk 22]
```diff
@@ -0,0 +1,481 @@
...
+    create_replication_config = DmsCreateReplicationConfigOperator(
+        task_id="create_replication_config",
+        replication_config_id=replication_id,
+        source_endpoint_arn=create_assets["source_endpoint_arn"],
+        target_endpoint_arn=create_assets["target_endpoint_arn"],
+        compute_config={
+            "MaxCapacityUnits": 4,
+            "MinCapacityUnits": 1,
+            "MultiAZ": False,
+            "ReplicationSubnetGroupId": "default",
```
Comment (ferruzzi) [inline providers/tests/system/amazon/aws/example_dms_serverless.py:360]: This test fails with the following stack trace:

```ERROR    airflow.task:taskinstance.py:3035 Task failed with exception
Traceback (most recent call last):
  File "/opt/airflow/providers/src/airflow/providers/amazon/aws/hooks/dms.py", line 281, in create_replication_config
    resp = self.conn.create_replication_config(
  File "/usr/local/lib/python3.9/site-packages/botocore/client.py", line 569, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.9/site-packages/botocore/client.py", line 1023, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidParameterValueException) when calling the CreateReplicationConfig operation: The provided value 'default' for field 'ReplicationSubnetGroupId' does not exist. Please verify the replication subnet group exists
```



Does it perhaps need a step to create (and later delete) a replication subnet group?  ([using this](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dms/client/create_replication_subnet_group.html))

Comment (ellisms) [reply providers/tests/system/amazon/aws/example_dms_serverless.py:360]: Odd, let me look into it.

Comment (ferruzzi) [reply providers/tests/system/amazon/aws/example_dms_serverless.py:360]: Any luck?

Comment (ellisms) [reply providers/tests/system/amazon/aws/example_dms_serverless.py:360]: No, I'm having a problem running the DAG in Breeze. It seems the xcoms aren't being stored from tasks such as `get_vpc_id`, so when subsequent tasks try to use those values, they are missing. Even though the logs say valid values are returned. Haven't been able to track down what's going on.

Comment (eladkal) [reply providers/tests/system/amazon/aws/example_dms_serverless.py:360]: Not stored at all or stored with different id?
The task in question is wrapped with task group but the xcom pull is missing the group id
https://stackoverflow.com/a/72137722/14624409

Comment (ellisms) [reply providers/tests/system/amazon/aws/example_dms_serverless.py:360]: I don't think it's being stored. When I view the `get_default_vpc_id` task in the Airflow UI, the Xcom table says "No Xcom". I confirmed the Xcom table in the databse is empty. The task log, however, shows the return value I expect:
`{"logger":"airflow.task.operators.airflow.decorators.python._PythonDecoratedOperator","timestamp":"2025-01-07T19:21:51.381738","event":"Done. Returned value was: vpc-067c44fffb613cda6","level":"info"}`

I didn't encounter this when I initially made the PR - everything worked. Has something changed?
